### ISSUE TITLE
[adjacent-overload-signatures] getters and setters are not overloads

### ISSUE THREAD

**Issue by Josh-Cena** [typescript-eslint/typescript-eslint#4576]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://typescript-eslint.io/docs/linting/troubleshooting) and my problem is not listed.

**Repro**

```ts
class A {
  #a = 1;
  #b = 2;

  get a(): number {
    return this.#a;
  }
  get b(): number {
    return this.#b;
  }
  set a(val: number) {
    this.#a = val;
  }
  set b(val: number) {
    this.#b = val;
  }
}
```

```JSON
{
  "rules": {
    "@typescript-eslint/adjacent-overload-signatures": "error"
  }
}
```

**Expected Result**

No errors. From my understanding, overload signatures are purely type-level, but getters and setters are separate runtime entities, hence they should not be seen as overload signatures. I also think this rule should differentiate itself from [`accessor-pairs`](https://eslint.org/docs/rules/accessor-pairs).

**Actual Result**

<img width="686" alt="image" src="https://user-images.githubusercontent.com/55398995/154835734-0987164c-8ebd-4eae-8ddd-e50a7140b827.png">

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `5.12.0` |
| `@typescript-eslint/parser`        | `5.12.0` |
| `TypeScript`                       | `4.5.5` |
| `ESLint`                           | `8.9.0` |
| `node`                             | `17.5.0` |

**Comment by bradzacher** [comment]
> I also think this rule should differentiate itself from [accessor-pairs](https://eslint.org/docs/rules/accessor-pairs).

[accessor-pairs](https://eslint.org/demo#eyJ0ZXh0IjoiLyogZXNsaW50IFwiYWNjZXNzb3ItcGFpcnNcIjogW1xuICAgIFwiZXJyb3JcIixcbl0gKi9cblxuY2xhc3MgQSB7XG4gICNhID0gMTtcbiAgI2IgPSAyO1xuXG4gIGdldCBhKCkge1xuICAgIHJldHVybiB0aGlzLiNhO1xuICB9XG4gIGdldCBiKCkge1xuICAgIHJldHVybiB0aGlzLiNiO1xuICB9XG4gIHNldCBhKHZhbCkge1xuICAgIHRoaXMuI2EgPSB2YWw7XG4gIH1cbiAgc2V0IGIodmFsKSB7XG4gICAgdGhpcy4jYiA9IHZhbDtcbiAgfVxufSIsIm9wdGlvbnMiOnsicGFyc2VyT3B0aW9ucyI6eyJlY21hVmVyc2lvbiI6ImxhdGVzdCIsInNvdXJjZVR5cGUiOiJtb2R1bGUiLCJlY21hRmVhdHVyZXMiOnt9fSwicnVsZXMiOnt9LCJlbnYiOnt9fX0=) wouldn't warn on this - it just warns about *missing* getters/setters.
OTOH [grouped-accessor-pairs](https://eslint.org/demo#eyJ0ZXh0IjoiLyogZXNsaW50IFwiZ3JvdXBlZC1hY2Nlc3Nvci1wYWlyc1wiOiBbXG4gICAgXCJlcnJvclwiLFxuXSAqL1xuXG5jbGFzcyBBIHtcbiAgI2EgPSAxO1xuICAjYiA9IDI7XG5cbiAgZ2V0IGEoKSB7XG4gICAgcmV0dXJuIHRoaXMuI2E7XG4gIH1cbiAgZ2V0IGIoKSB7XG4gICAgcmV0dXJuIHRoaXMuI2I7XG4gIH1cbiAgc2V0IGEodmFsKSB7XG4gICAgdGhpcy4jYSA9IHZhbDtcbiAgfVxuICBzZXQgYih2YWwpIHtcbiAgICB0aGlzLiNiID0gdmFsO1xuICB9XG59Iiwib3B0aW9ucyI6eyJwYXJzZXJPcHRpb25zIjp7ImVjbWFWZXJzaW9uIjoibGF0ZXN0Iiwic291cmNlVHlwZSI6Im1vZHVsZSIsImVjbWFGZWF0dXJlcyI6e319LCJydWxlcyI6e30sImVudiI6e319fQ==) will warn about this case.

-----

I 100% understand where you're coming from. If this were a new rule I would strongly agree with the stance!

At this point, however, the rule has been implemented with this exact logic since its implementation (4 and a half years ago) or back over 5 years including the TSLint rule is was born from.
I think that how it works must just be how almost all users expect the rule to function!

**Comment by Josh-Cena** [comment]
> OTOH [grouped-accessor-pairs](https://eslint.org/demo#eyJ0ZXh0IjoiLyogZXNsaW50IFwiZ3JvdXBlZC1hY2Nlc3Nvci1wYWlyc1wiOiBbXG4gICAgXCJlcnJvclwiLFxuXSAqL1xuXG5jbGFzcyBBIHtcbiAgI2EgPSAxO1xuICAjYiA9IDI7XG5cbiAgZ2V0IGEoKSB7XG4gICAgcmV0dXJuIHRoaXMuI2E7XG4gIH1cbiAgZ2V0IGIoKSB7XG4gICAgcmV0dXJuIHRoaXMuI2I7XG4gIH1cbiAgc2V0IGEodmFsKSB7XG4gICAgdGhpcy4jYSA9IHZhbDtcbiAgfVxuICBzZXQgYih2YWwpIHtcbiAgICB0aGlzLiNiID0gdmFsO1xuICB9XG59Iiwib3B0aW9ucyI6eyJwYXJzZXJPcHRpb25zIjp7ImVjbWFWZXJzaW9uIjoibGF0ZXN0Iiwic291cmNlVHlwZSI6Im1vZHVsZSIsImVjbWFGZWF0dXJlcyI6e319LCJydWxlcyI6e30sImVudiI6e319fQ==) will warn about this case.

My bad, meant `grouped-accessor-pairs` :D

> I think that how it works must just be how almost all users expect the rule to function

Accessors fall in the gray area where it's neither documented nor tested. I understand things like Hyrum's law, but the facts are:

- This deviates from the expectation of what "overload signatures" mean.
- This means less errors, not more, so it is by no means breaking.
- Removal of this behavior can be easily compensated through enabling `grouped-accessor-pairs`.

I still do believe that it's a bug and we need to fix it.

**Comment by bradzacher** [comment]
> but the facts are:
> - This deviates from the expectation of what "overload signatures" mean.

My argument is that most users don't see this as a deviation. If they did then we would have gotten an issue earlier! Either that, or not enough people use this rule to notice. Hard to say as we can't collect metrics.

> - Fixing this means less errors, not more, so it is by no means breaking.
> - Removal of this behavior can be easily compensated through enabling grouped-accessor-pairs.

Users would *lose* coverage that they may have come to rely on, and will be required to change their config to turn on a new rule.
Which is why I would class the change as a breaking change.

Again - if this issue was a few years ago, I wouldn't be making that distinction at all! But at this point it's more or less just an undocumented part of the rule.

**Comment by Josh-Cena** [comment]
So, the plan would be to have a default-true option to turn this behavior off, and in the next major version we remove this option and just ignore accessors for this rule?

**Comment by bradzacher** [comment]
> in the next major version we remove this option and just ignore accessors for this rule

Probably just leave the option as is. As mentioned - everyone is happy enough with the behaviour so no sense disrupting people without a really good reason

**Comment by Josh-Cena** [comment]
Well... this certainly seems like a bug to me, and putting myself into the shoes a project maintainer, I don't feel like we should ever carry this tech debt forever for "backward compatibility" ㄟ(▔,▔)ㄏ But as long as there's an option, it should be good!

**Comment by yokuze** [comment]
FWIW, I agree with @Josh-Cena that this is a bug. It's really confusing that this error reports violations for getters/setters because getters/setters have nothing to do with overload signatures.

> My argument is that most users don't see this as a deviation. If they did then we would have gotten an issue earlier! 

In our case, we did not notice this deviation until now because we've always had `grouped-accessor-pairs` enabled. I was re-configuring our linting rules today to demonstrate some different options, noticed it, and am now reporting it. Just adding it as a data point against the "we would have gotten an issue earlier" argument.

When we disable `grouped-accessor-pairs` and enable `member-ordering` with its default configuration, this rule throws an error that's unrelated to the rule's documented purpose. I would not want to rely on that unusual behavior to enforce a linting standard, especially since the subsequent error message (i.e. "All ${PROP_NAME} signatures should be adjacent") sends you looking in the wrong direction to fix it.

**Comment by yovanoc** [comment]
Currently annoyed from this thing with default config sadly

**Comment by Josh-Cena** [comment]
I'm taking a judgment call and relabeling this as a https://github.com/typescript-eslint/typescript-eslint/labels/bug that should just be fixed by default, since it doesn't result in more errors and people can easily revert to the previous behavior by turning on `grouped-accessor-pairs`.

**Comment by bradzacher** [comment]
[Here is a list of almost 3 thousand repos thatuse the rule and don't have an issue with the accessor grouping behaviour](https://sourcegraph.com/search?q=context:global+/adjacent-overload-signatures%5B%22%27%5D:%5Cs*%5C%5B%3F%5Cs*%28%5B%27%22%5Derror%5B%27%22%5D%7C%5B%27%22%5Dwarn%5B%27%22%5D%7C1%7C2%29/+-repo:%5Egithub%5C.com/pnp/sp-dev-fx-webparts%24+count:all&patternType=keyword&sm=0).

I STRONGLY disagree that it's a bug. As I mentioned earlier - if this were a newish rule I'd be right there with you. But given the age of this behaviour (over 5 years) and the fact that nobody has reported it before you - that suggests that it's what most people expect.
From that I would not say "this is a problem with the rule" and instead say "this is working as intended". Hence it's an enhancement to change the behaviour and not a bug.

> and people can easily revert to the previous behavior by turning on grouped-accessor-pairs

That's not an "easily" thing though.
This change is invisible to users - it turns an active report into a non-report.
Users can't see that this change ever occurred **_unless they specifically read the patch notes_** -- which we know most people don't do.

So you're taking away functionality and assigning it to a different rule and users have to manually change their config in order to regain that functionality.
But given the removal is invisible to users - they may just lose coverage without ever knowing it.

Removing this by default is definitely a breaking change.

---

This is why I suggested adding an option. We can add the option to disable the behaviour so that the small minority that don't want this behaviour can have it without impacting other users.
Then in a major we can flip the default, then either remove it and advise `grouped-accessors` or just leave it in.

**Comment by Josh-Cena** [comment]
OK, I concur.
